### ISSUE TITLE
Question: Best way to load Snapin Vendor (not a module)

### ISSUE THREAD

**Issue by Tonic8** [Badgerati/Pode#204]
Hello 
Thanks for Pode :) wonderful tool!

i try to set a API rest system that need a snap loading (not a module).

here a simple example

```
Server {
    listen localhost:8080 http

    route get '/api/test' {
		param($s)

		add-pssnapin citrix.broker.admin.v2

		$result = get-brokerdesktop -machinename $s.Parameters['MachineName'] -Adminaddress server1

		json @{
			'machinename' = $s.Parameters['MachineName'];
			'CatalogName' = $result.'CatalogName';
			'PowerState' = $result.'PowerState';
		}
    }
}
```

when i try, the first time, it work, but requesting a second time it fail.

+ there is a pb each time we call the route, it will load the snapin, not very optimised

not yet tested function + dot source the script, but we fall in the same pb (loading unloading the snapin each time it is called.

thanks
Tonic8

**Comment by Badgerati** [comment]
Hey @Tonic8 ,
Glad you're enjoying Pode! :)

Snapins are something I've not had to use much, but like modules I know there is a method to cleanly import them onto each runspace - without having to call `Add-PSSnapIn` each time. This way of doing it is something I'll have to test within Pode, and implement it as an extension of the `import` function.

For your current issue, one way you might be able to do this without have to load/unload every time is to try using the [`import`](https://badgerati.github.io/Pode/Tutorials/ImportingModules/) function; placing the `Add-SnapIn citrix.broker.admin.v2` into a module and reference that. This way it'll only ever be called once for each runspace Pode uses.

Thanks,
Badgerati

**Comment by Tonic8** [comment]
Hello

thanks for the trick (and the fast response) it work at least there is no fail while using it,

```
function import-CitrixSnapin {
   Add-SnapIn citrix.broker.admin.v2
   write-host "successfully loaded"
}
```

```
Server {
    listen localhost:8080 http
    import c:\tet\pode\ps_module\CitrixSnapin\CitrixSnapin.ps1
    route get '/api/test' {
		param($s)

		import-CitrixSnapin

		$result = get-brokerdesktop -machinename $s.Parameters['MachineName'] -Adminaddress server1

		json @{
			'machinename' = $s.Parameters['MachineName'];
			'CatalogName' = $result.'CatalogName';
			'PowerState' = $result.'PowerState';
		}
    }
}

```

but it relaunch it each time it is called :), so i think i did it wrong?

Tonic8

**Comment by Badgerati** [comment]
Hey,

In the module, you should be able to just have `Add-SnapIn citrix.broker.admin.v2` without it being in a function; when the module is imported that line will be run automatically :)

**Comment by Tonic8** [comment]
Hello

tested and not working :(, 2 things:
- i tried this way before and was not successfull, i spotted an issue on this, so i clean everything and used the code you just provided with a single modification 'write-host "successfully loaded"' in the module ps1 file. When i load the server.ps1 file, the line appear 2 or more time , and that's before i try to request

```
C:\workspace\pode\test> .\server.ps1
successfully loaded
successfully loaded
successfully loaded
successfully loaded
successfully loaded
Listening on the following 1 endpoint(s) [1 thread(s)]:
            - http://localhost:8080/
successfully loaded
successfully loaded
```

- when i try to do the request, on server side i got:

```
C:\workspace\pode\test> .\server.ps1
Listening on the following 1 endpoint(s) [1 thread(s)]: 
		http://1ocalhost:8080/ 
Cannot index into a null array. 
At C:\workspace\pode\test\server.ps1:16 char:3 
+           $result = get-brokerdesktop -machinename $s.parameters['Machi ... 
+           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
	+ CategoryInfo			: Invalidoperation: (:) [] , RuntimeException 
	+ FullyQua1ifiedErrorId : Nul1Array 
```

if i do it "my" way (calling the function inside the route block) it worked :)

Tonic8

**Comment by Badgerati** [comment]
Hey,

Yes that's right, after the "listening" message appears the last 2 "success" messages are the final runspaces being created.

The error will be because of the `route` path being `/api/test`, so the `$s.Parameters` array will be null - throwing that error. Changing that to something like `/api/test/:MachineName` should stop that null error from happening.

Badgerati

**Comment by LLIT** [comment]
How about:

```
$SnapinName = "citrix.broker.admin.v2"
if (!(Get-PSSnapin -Name $SnapinName)) {
    Write-Host "Adding Snapin: $($SnapinName)"
    Add-PSSnapIn $SnapinName -ErrorAction Stop
}
```

And then just place it in every Route
I do this with Modules

**Comment by Tonic8** [comment]
Hello LLIT the problem here is that each time you call a route you load the snapin, when the operation is finnish and the route it unload the snapin because the context of the route is droped.

when you load the module outside of the route with the "import c:\tet\pode\ps_module\CitrixSnapin\CitrixSnapin.ps1" it's loaded once the server is loaded at the beginning. but not each time you execute a route, is not loaded.

but for now is ok, tested it work pretty well. just the action in memory if the snapin is heavy, can affect the performance. in my case hopefully is ok.

**Comment by Badgerati** [comment]
Hey,

Loading the SnapIn within the module should import it onto each runspace pool, and runspace, that Pode uses - though there's a chance that SnapIns work a little differently. Routes each run within a runspace session that will have had the module run + have the SnapIn loaded via the module trick; and in theory LLIT's idea should work, which makes me wonder if there really is something weird about SnapIns 🤔.

For the next release (v0.28.0), and currently in the `develop` branch, I've added support on the `import` function to utilise the actual PSSnapIn functions on the runspace pools to import SnapIns directly. So you will be able to do:

```powershell
import -snapin 'citrix.broker.admin.v2'
```

If you want to try using the develop branch, this should hopefully work for you, instead of trying the module trick - if not, is there anywhere you might be able to place your scripts?

**Comment by Badgerati** [comment]
Hey @Tonic8,

The SnapIn support on the `import` function, from the above comment, was released in 0.28.0 - hope it does the trick for you :)

**Comment by Tonic8** [comment]
:) thanks, will proceed to the test ASAP.
